Classes/common/touchdb/TD_Database.m

```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
+    if (result && key) {
+        [queue inDatabase:^(FMDatabase* db) {
+            result = [db setKey:key];
+            if (!result) {
+                CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"Key to cipher database not set");
+            } else {
+                result = (sqlite3_exec(db.sqliteHandle, "SELECT count(*) FROM sqlite_master;",
+                                       NULL, NULL, NULL) == SQLITE_OK);
+                if (!result) {
+                    CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"Wrong key");
+                }
+            }
+        }];
+    }
 
     // Register CouchDB-compatible JSON collation functions:
-    [_fmdbQueue inDatabase:^(FMDatabase* db) {
-        sqlite3_create_collation(db.sqliteHandle, "JSON", SQLITE_UTF8, kTDCollateJSON_Unicode,
-                                 TDCollateJSON);
-        sqlite3_create_collation(db.sqliteHandle, "JSON_RAW", SQLITE_UTF8, kTDCollateJSON_Raw,
-                                 TDCollateJSON);
-        sqlite3_create_collation(db.sqliteHandle, "JSON_ASCII", SQLITE_UTF8, kTDCollateJSON_ASCII,
-                                 TDCollateJSON);
-        sqlite3_create_collation(db.sqliteHandle, "REVID", SQLITE_UTF8, NULL, TDCollateRevIDs);
-    }];
+    if (result) {
+        [queue inDatabase:^(FMDatabase* db) {
+          sqlite3_create_collation(db.sqliteHandle, "JSON", SQLITE_UTF8, kTDCollateJSON_Unicode,
+                                   TDCollateJSON);
+          sqlite3_create_collation(db.sqliteHandle, "JSON_RAW", SQLITE_UTF8, kTDCollateJSON_Raw,
+                                   TDCollateJSON);
+          sqlite3_create_collation(db.sqliteHandle, "JSON_ASCII", SQLITE_UTF8, kTDCollateJSON_ASCII,
+                                   TDCollateJSON);
+          sqlite3_create_collation(db.sqliteHandle, "REVID", SQLITE_UTF8, NULL, TDCollateRevIDs);
+        }];
+    }
 
-    __block BOOL result = YES;
-    __weak TD_Database* weakSelf = self;
-    [_fmdbQueue inDatabase:^(FMDatabase* db) {
-        // Stuff we need to initialize every time the database opens:
-        TD_Database* strongSelf = weakSelf;
-        if (![strongSelf initialize:@"PRAGMA foreign_keys = ON;" inDatabase:db]) result = NO;
-    }];
-    if (!result) {
-        return NO;
+    // Stuff we need to initialize every time the database opens:
+    if (result) {
+        __weak TD_Database* weakSelf = self;
+        [queue inDatabase:^(FMDatabase* db) {
+          TD_Database* strongSelf = weakSelf;
+          if (!strongSelf || ![strongSelf initialize:@"PRAGMA foreign_keys = ON;" inDatabase:db]) {
+              result = NO;
+          }
+        }];
+    }
+    
+    // Assign properties (if everything was OK)
+    if (result) {
+        _fmdbQueue = queue;
+        _keyProviderToOpenDB = provider;
+    } else if (queue) {
+        [queue close];
     }
 
-    return YES;
+    return result;
+}
+
+// callers: many things
+- (BOOL)isOpen
+{
+    return _open;
+}
+
+// callers: many things
```

Comment:
Yes but it is a public method, potentially it can be called from many places. Right now only `openWithEncryptionKeyProvider:` calls it but I think we can need it later on so I made it public.